### ISSUE TITLE
Can we force Z3 to explore certain candidate space first?

### ISSUE THREAD

**Issue by anubhavnidhi** [Z3Prover/z3#2047]
I had a question on smt-solvers (like Z3) and was wondering if you know any Z3-tactic which can help me achieve my objective.

I want to know whether it is possible to force Z3 to explore some variables before exploring other variables.

For examples, I have a scenario where my MaxSMT problem has the following hard constraints 

    X1 + X4 >=3
    X2 + 7  >=3
    X3 + 8  >=3
    

And the soft-constraint is 

    X4 == 0

Here I want to force the smt-solver to first explore candidate space with different values for variable X1. (I believe by default Z3 will randomly explore different values of X1, X2, X3, X4)

So my question is - is there any tactic in Z3 that allows me to tell the Z3-solver what candidate space (set of variables) should it try exploring first?

**Comment by NikolajBjorner** [comment]
I am not sure how this question translates into a usable tactic. None of the available tactics fall into this category. Note that Z3 isn't based on a Constraint Satisfaction engine. Thus, it doesn't work by arc-consistency and fixing variables. So the notion of exploring different values for variables doesn't follow the Constraint Programming, CP, terminology. Values to arithmetic variables are explored by pivoting in a Simplex tableau and in some cases for finite domains tactics translate integers to bit-vectors and use propositional SAT engines. 
So in summary, even though the question isn't stated in a precise way, I would say "no" there isn't a tactic that fits your question.

**Comment by pcarbonn** [comment]
One option is to use the [cube](https://z3prover.github.io/api/html/classz3py_1_1_solver.html#a64a676332f400ccf34ee240f7f7f964c) method of the Solver class. It allows you to specify the variables to use for case split.
